File: packages/cordova-plugin-notificare-authentication/src/android/NotificareAuthenticationPlugin.kt
package re.notifica.authentication.cordova

import android.content.Intent
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaArgs
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.authentication.ktx.authentication
import re.notifica.authentication.models.NotificareUser
import re.notifica.authentication.models.NotificareUserPreference
import re.notifica.authentication.models.NotificareUserSegment
import re.notifica.internal.NotificareLogger

class NotificareAuthenticationPlugin : CordovaPlugin() {

    override fun pluginInitialize() {
        val intent = cordova.activity.intent
        if (intent != null) onNewIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        val passwordResetToken = Notificare.authentication().parsePasswordResetToken(intent)
        if (passwordResetToken != null) {
            NotificareAuthenticationPluginEventBroker.dispatchEvent("password_reset_token_received", passwordResetToken)
            return
        }

        val validateUserToken = Notificare.authentication().parseValidateUserToken(intent)
        if (validateUserToken != null) {
            NotificareAuthenticationPluginEventBroker.dispatchEvent("validate_user_token_received", validateUserToken)
            return
        }
    }

    override fun execute(action: String, args: CordovaArgs, callback: CallbackContext): Boolean {
        when (action) {
            "isLoggedIn" -> isLoggedIn(args, callback)
            "login" -> login(args, callback)
            "logout" -> logout(args, callback)
            "fetchUserDetails" -> fetchUserDetails(args, callback)
            "changePassword" -> changePassword(args, callback)
            "generatePushEmailAddress" -> generatePushEmailAddress(args, callback)
            "createAccount" -> createAccount(args, callback)
            "validateUser" -> validateUser(args, callback)
            "sendPasswordReset" -> sendPasswordReset(args, callback)
            "resetPassword" -> resetPassword(args, callback)
            "fetchUserPreferences" -> fetchUserPreferences(args, callback)
            "fetchUserSegments" -> fetchUserSegments(args, callback)
            "addUserSegment" -> addUserSegment(args, callback)
            "removeUserSegment" -> removeUserSegment(args, callback)
            "addUserSegmentToPreference" -> addUserSegmentToPreference(args, callback)
            "removeUserSegmentFromPreference" -> removeUserSegmentFromPreference(args, callback)

            // Event broker
            "registerListener" -> registerListener(args, callback)

            else -> {
                callback.error("No implementation for action '$action'.")
                return false
            }
        }

        return true
    }

    // region Notificare Authentication

    private fun isLoggedIn(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        callback.success(Notificare.authentication().isLoggedIn)
    }

    private fun login(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val email = args.getString(0)
        val password = args.getString(1)

        Notificare.authentication().login(email, password, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun logout(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Notificare.authentication().logout(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun fetchUserDetails(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Notificare.authentication().fetchUserDetails(object : NotificareCallback<NotificareUser> {
            override fun onSuccess(result: NotificareUser) {
                try {
                    callback.success(result.toJson())
                } catch (e: Exception) {
                    callback.error(e.message)
                }
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun changePassword(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val password = args.getString(0)

        Notificare.authentication().changePassword(password, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun generatePushEmailAddress(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Notificare.authentication().generatePushEmailAddress(object : NotificareCallback<NotificareUser> {
            override fun onSuccess(result: NotificareUser) {
                try {
                    callback.success(result.toJson())
                } catch (e: Exception) {
                    callback.error(e.message)
                }
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun createAccount(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val email = args.getString(0)
        val password = args.getString(1)
        val name = args.getString(2)

        Notificare.authentication().createAccount(email, password, name, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun validateUser(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val token = args.getString(0)

        Notificare.authentication().validateUser(token, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun sendPasswordReset(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val email = args.getString(0)

        Notificare.authentication().sendPasswordReset(email, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun resetPassword(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val password = args.getString(0)
        val token = args.getString(1)

        Notificare.authentication().resetPassword(password, token, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun fetchUserPreferences(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Notificare.authentication().fetchUserPreferences(object : NotificareCallback<List<NotificareUserPreference>> {
            override fun onSuccess(result: List<NotificareUserPreference>) {
                try {
                    val json = JSONArray()
                    result.forEach { json.put(it.toJson()) }

                    callback.success(json)
                } catch (e: Exception) {
                    callback.error(e.message)
                }
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun fetchUserSegments(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        Notificare.authentication().fetchUserSegments(object : NotificareCallback<List<NotificareUserSegment>> {
            override fun onSuccess(result: List<NotificareUserSegment>) {
                try {
                    val json = JSONArray()
                    result.forEach { json.put(it.toJson()) }

                    callback.success(json)
                } catch (e: Exception) {
                    callback.error(e.message)
                }
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun addUserSegment(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val segment = try {
            NotificareUserSegment.fromJson(args.getJSONObject(0))
        } catch (e: Exception) {
            callback.error(e.message)
            return
        }

        Notificare.authentication().addUserSegment(segment, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun removeUserSegment(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val segment = try {
            NotificareUserSegment.fromJson(args.getJSONObject(0))
        } catch (e: Exception) {
            callback.error(e.message)
            return
        }

        Notificare.authentication().removeUserSegment(segment, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                callback.void()
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    private fun addUserSegmentToPreference(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val preference = try {
            NotificareUserPreference.fromJson(args.getJSONObject(0))
        } catch (e: Exception) {
            callback.error(e.message)
            return
        }

        try {
            val segment = NotificareUserSegment.fromJson(args.getJSONObject(1))
            Notificare.authentication()
                .addUserSegmentToPreference(segment, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        callback.void()
                    }

                    override fun onFailure(e: Exception) {
                        callback.error(e.message)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserSegment.", e)
        }

        try {
            val option = NotificareUserPreference.Option.fromJson(args.getJSONObject(1))
            Notificare.authentication()
                .addUserSegmentToPreference(option, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        callback.void()
                    }

                    override fun onFailure(e: Exception) {
                        callback.error(e.message)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserPreference.Option.", e)
        }

        callback.error("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }

    private fun removeUserSegmentFromPreference(
        @Suppress("UNUSED_PARAMETER") args: CordovaArgs,
        callback: CallbackContext
    ) {
        val preference = try {
            NotificareUserPreference.fromJson(args.getJSONObject(0))
        } catch (e: Exception) {
            callback.error(e.message)
            return
        }

        try {
            val segment = NotificareUserSegment.fromJson(args.getJSONObject(1))
            Notificare.authentication()
                .removeUserSegmentFromPreference(segment, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        callback.void()
                    }

                    override fun onFailure(e: Exception) {
                        callback.error(e.message)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserSegment.", e)
        }

        try {
            val option = NotificareUserPreference.Option.fromJson(args.getJSONObject(1))
            Notificare.authentication()
                .removeUserSegmentFromPreference(option, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        callback.void()
                    }

                    override fun onFailure(e: Exception) {
                        callback.error(e.message)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserPreference.Option.", e)
        }

        callback.error("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }

    // endregion

    private fun registerListener(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        NotificareAuthenticationPluginEventBroker.setup(object : NotificareAuthenticationPluginEventBroker.Consumer {
            override fun onEvent(event: NotificareAuthenticationPluginEventBroker.Event) {
                val payload = JSONObject()
                payload.put("name", event.name)
                when (event.payload) {
                    null -> {} // Skip encoding null payloads.
                    is Boolean -> payload.put("data", event.payload)
                    is Int -> payload.put("data", event.payload)
                    is Float -> payload.put("data", event.payload)
                    is Double -> payload.put("data", event.payload)
                    is String -> payload.put("data", event.payload)
                    is JSONObject -> payload.put("data", event.payload)
                    is JSONArray -> payload.put("data", event.payload)
                    else -> throw IllegalArgumentException("Unsupported event payload of type '${event.payload::class.java.simpleName}'.")
                }

                val result = PluginResult(PluginResult.Status.OK, payload)
                result.keepCallback = true

                callback.sendPluginResult(result)
            }
        })
    }
}

private fun CallbackContext.void() {
    sendPluginResult(PluginResult(PluginResult.Status.OK, null as String?))
}

private fun CallbackContext.success(b: Boolean) {
    sendPluginResult(PluginResult(PluginResult.Status.OK, b))
}

File: packages/cordova-plugin-notificare-authentication/src/ios/NotificareAuthenticationPlugin.swift
import NotificareKit
import NotificareAuthenticationKit

@objc(NotificareAuthenticationPlugin)
class NotificareAuthenticationPlugin : CDVPlugin {

    override func pluginInitialize() {
        super.pluginInitialize()

        _ = NotificareSwizzler.addInterceptor(self)
    }

    @objc func registerListener(_ command: CDVInvokedUrlCommand) {
        NotificareAuthenticationPluginEventBroker.startListening(settings: commandDelegate.settings) { event in
            var payload: [String: Any] = [
                "name": event.name,
            ]

            if let data = event.payload {
                payload["data"] = data
            }

            let result = CDVPluginResult(status: .ok, messageAs: payload)
            result!.keepCallback = true

            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    // MARK: - Notificare Authentication

    @objc func isLoggedIn(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: Notificare.shared.authentication().isLoggedIn)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func login(_ command: CDVInvokedUrlCommand) {
        let email = command.argument(at: 0) as! String
        let password = command.argument(at: 1) as! String

        Notificare.shared.authentication().login(email: email, password: password) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func logout(_ command: CDVInvokedUrlCommand) {
        Notificare.shared.authentication().logout { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func fetchUserDetails(_ command: CDVInvokedUrlCommand) {
        Notificare.shared.authentication().fetchUserDetails { result in
            switch result {
            case let .success(user):
                do {
                    let json = try user.toJson()

                    let result = CDVPluginResult(status: .ok, messageAs: json)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } catch {
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func changePassword(_ command: CDVInvokedUrlCommand) {
        let password = command.argument(at: 0) as! String

        Notificare.shared.authentication().changePassword(password) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func generatePushEmailAddress(_ command: CDVInvokedUrlCommand) {
        Notificare.shared.authentication().generatePushEmailAddress { result in
            switch result {
            case let .success(user):
                do {
                    let json = try user.toJson()

                    let result = CDVPluginResult(status: .ok, messageAs: json)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } catch {
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func createAccount(_ command: CDVInvokedUrlCommand) {
        let email = command.argument(at: 0) as! String
        let password = command.argument(at: 1) as! String
        let name = command.argument(at: 2) as! String?

        Notificare.shared.authentication().createAccount(email: email, password: password, name: name) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func validateUser(_ command: CDVInvokedUrlCommand) {
        let token = command.argument(at: 0) as! String

        Notificare.shared.authentication().validateUser(token: token) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func sendPasswordReset(_ command: CDVInvokedUrlCommand) {
        let email = command.argument(at: 0) as! String

        Notificare.shared.authentication().sendPasswordReset(email: email) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func resetPassword(_ command: CDVInvokedUrlCommand) {
        let password = command.argument(at: 0) as! String
        let token = command.argument(at: 1) as! String

        Notificare.shared.authentication().resetPassword(password, token: token) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func fetchUserPreferences(_ command: CDVInvokedUrlCommand) {
        Notificare.shared.authentication().fetchUserPreferences { result in
            switch result {
            case let .success(preferences):
                do {
                    let json = try preferences.map { try $0.toJson() }

                    let result = CDVPluginResult(status: .ok, messageAs: json)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } catch {
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func fetchUserSegments(_ command: CDVInvokedUrlCommand) {
        Notificare.shared.authentication().fetchUserSegments { result in
            switch result {
            case let .success(segments):
                do {
                    let json = try segments.map { try $0.toJson() }

                    let result = CDVPluginResult(status: .ok, messageAs: json)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } catch {
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func addUserSegment(_ command: CDVInvokedUrlCommand) {
        let segment: NotificareUserSegment

        do {
            let json = command.argument(at: 0) as! [String: Any]
            segment = try NotificareUserSegment.fromJson(json: json)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        Notificare.shared.authentication().addUserSegment(segment) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }

    @objc func removeUserSegment(_ command: CDVInvokedUrlCommand) {
        let segment: NotificareUserSegment

        do {
            let json = command.argument(at: 0) as! [String: Any]
            segment = try NotificareUserSegment.fromJson(json: json)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        Notificare.shared.authentication().removeUserSegment(segment) { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: .ok)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }


    @objc func addUserSegmentToPreference(_ command: CDVInvokedUrlCommand) {
        let preference: NotificareUserPreference

        do {
            let json = command.argument(at: 0) as! [String: Any]
            preference = try NotificareUserPreference.fromJson(json: json)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        do {
            let json = command.argument(at: 1) as! [String: Any]
            let segment = try NotificareUserSegment.fromJson(json: json)
            Notificare.shared.authentication().addUserSegmentToPreference(segment, to: preference) { result in
                switch result {
                case .success:
                    let result = CDVPluginResult(status: .ok)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                case let .failure(error):
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            }

            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserSegment.", error: error)
        }

        do {
            let json = command.argument(at: 1) as! [String: Any]
            let option = try NotificareUserPreference.Option.fromJson(json: json)
            Notificare.shared.authentication().addUserSegmentToPreference(option: option, to: preference) { result in
                switch result {
                case .success:
                    let result = CDVPluginResult(status: .ok)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                case let .failure(error):
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            }

            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserPreference.Option.", error: error)
        }

        let result = CDVPluginResult(status: .error, messageAs: "To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func removeUserSegmentFromPreference(_ command: CDVInvokedUrlCommand) {
        let preference: NotificareUserPreference

        do {
            let json = command.argument(at: 0) as! [String: Any]
            preference = try NotificareUserPreference.fromJson(json: json)
        } catch {
            let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
            return
        }

        do {
            let json = command.argument(at: 1) as! [String: Any]
            let segment = try NotificareUserSegment.fromJson(json: json)
            Notificare.shared.authentication().removeUserSegmentFromPreference(segment, from: preference) { result in
                switch result {
                case .success:
                    let result = CDVPluginResult(status: .ok)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                case let .failure(error):
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            }

            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserSegment.", error: error)
        }

        do {
            let json = command.argument(at: 1) as! [String: Any]
            let option = try NotificareUserPreference.Option.fromJson(json: json)
            Notificare.shared.authentication().removeUserSegmentFromPreference(option: option, from: preference) { result in
                switch result {
                case .success:
                    let result = CDVPluginResult(status: .ok)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                case let .failure(error):
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            }

            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserPreference.Option.", error: error)
        }

        let result = CDVPluginResult(status: .error, messageAs: "To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
}

extension NotificareAuthenticationPlugin: NotificareAppDelegateInterceptor {
    func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
        guard let url = userActivity.webpageURL else {
            return false
        }

        if let token = Notificare.shared.authentication().parsePasswordResetToken(url) {
            NotificareAuthenticationPluginEventBroker.dispatchEvent(name: "password_reset_token_received", payload: token)
            return true
        }

        if let token = Notificare.shared.authentication().parseValidateUserToken(url) {
            NotificareAuthenticationPluginEventBroker.dispatchEvent(name: "validate_user_token_received", payload: token)
            return false
        }

        return false
    }
}
